File: sdks/php/src/Api/Metrics.php
<?php

declare(strict_types=1);

namespace Apollo\Deploy\Signal\Sdk\Api;

use Apollo\Deploy\Signal\Sdk\SDKError;
use Apollo\Deploy\Signal\Sdk\SSEEvent;
use Apollo\Deploy\Signal\Sdk\Transport;
use Apollo\Deploy\Signal\Sdk\AdvisorReportResponse;
use Apollo\Deploy\Signal\Sdk\EmailPerformanceResponse;
use Apollo\Deploy\Signal\Sdk\EmailReadScrollAnalyticsResponse;
use Apollo\Deploy\Signal\Sdk\GetMetricsAdvisorQuery;
use Apollo\Deploy\Signal\Sdk\GetProjectMetricsSummaryQuery;
use Apollo\Deploy\Signal\Sdk\GetProjectMetricsTimelineQuery;
use Apollo\Deploy\Signal\Sdk\GetTopicPerformanceQuery;
use Apollo\Deploy\Signal\Sdk\ProjectMetricsSummaryResponse;
use Apollo\Deploy\Signal\Sdk\ProjectMetricsTimelineResponse;
use Apollo\Deploy\Signal\Sdk\TopicPerformanceResponse;
use Psr\Http\Message\ResponseInterface;

/** MetricsAPI domain operations. */
class Metrics
{
    public function __construct(private Transport $transport)
    {
    }

    /**
     * @param string $id
     * @param string|null $window Time window.
     * @param string|null $projectId Required for session auth; ignored for API key auth.
     * @return TopicPerformanceResponse
     * @throws SDKError
     */
    public function getTopicPerformance(
        string $id,
        ?string $window = null,
        ?string $projectId = null,
    ): TopicPerformanceResponse
    {
        $path = '/v1/metrics/topics/${id}';
        $path = str_replace('${id}', rawurlencode(Transport::wireValue($id)), $path);

        $query = [];
        if ($window !== null) {
            $query['window'] = $window;
        }
        if ($projectId !== null) {
            $query['projectId'] = $projectId;
        }

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(TopicPerformanceResponse::class, 'fromArray')) {
            return TopicPerformanceResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $id
     * @return EmailPerformanceResponse
     * @throws SDKError
     */
    public function getEmailPerformance(
        string $id,
    ): EmailPerformanceResponse
    {
        $path = '/v1/metrics/emails/${id}';
        $path = str_replace('${id}', rawurlencode(Transport::wireValue($id)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(EmailPerformanceResponse::class, 'fromArray')) {
            return EmailPerformanceResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $id
     * @return EmailReadScrollAnalyticsResponse
     * @throws SDKError
     */
    public function getEmailEngagement(
        string $id,
    ): EmailReadScrollAnalyticsResponse
    {
        $path = '/v1/metrics/emails/${id}/engagement';
        $path = str_replace('${id}', rawurlencode(Transport::wireValue($id)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(EmailReadScrollAnalyticsResponse::class, 'fromArray')) {
            return EmailReadScrollAnalyticsResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string|null $window Metrics time window.
     * @param string|null $projectId Required for session auth; ignored for API key auth.
     * @return ProjectMetricsSummaryResponse
     * @throws SDKError
     */
    public function getProjectMetricsSummary(
        ?string $window = null,
        ?string $projectId = null,
    ): ProjectMetricsSummaryResponse
    {
        $path = '/v1/metrics/summary';

        $query = [];
        if ($window !== null) {
            $query['window'] = $window;
        }
        if ($projectId !== null) {
            $query['projectId'] = $projectId;
        }

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(ProjectMetricsSummaryResponse::class, 'fromArray')) {
            return ProjectMetricsSummaryResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string|null $window Metrics time window.
     * @param string|null $granularity `hour` or `day`.
     * @param string|null $format `compact` or `detailed`.
     * @param string|null $projectId Required for session auth; ignored for API key auth.
     * @return ProjectMetricsTimelineResponse
     * @throws SDKError
     */
    public function getProjectMetricsTimeline(
        ?string $window = null,
        ?string $granularity = null,
        ?string $format = null,
        ?string $projectId = null,
    ): ProjectMetricsTimelineResponse
    {
        $path = '/v1/metrics/timeline';

        $query = [];
        if ($window !== null) {
            $query['window'] = $window;
        }
        if ($granularity !== null) {
            $query['granularity'] = $granularity;
        }
        if ($format !== null) {
            $query['format'] = $format;
        }
        if ($projectId !== null) {
            $query['projectId'] = $projectId;
        }

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(ProjectMetricsTimelineResponse::class, 'fromArray')) {
            return ProjectMetricsTimelineResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string|null $projectId Required for session auth; ignored for API key auth.
     * @return AdvisorReportResponse
     * @throws SDKError
     */
    public function getMetricsAdvisor(
        ?string $projectId = null,
    ): AdvisorReportResponse
    {
        $path = '/v1/metrics/advisor';

        $query = [];
        if ($projectId !== null) {
            $query['projectId'] = $projectId;
        }

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(AdvisorReportResponse::class, 'fromArray')) {
            return AdvisorReportResponse::fromArray($data);
        }
        return $data;
    }

    private function decodeResponse(ResponseInterface $response): mixed
    {
        $contents = $this->transport->readResponseBody($response);
        if (trim($contents) === '') {
            $requestId = $response->getHeaderLine('x-request-id')
                ?: $response->getHeaderLine('request-id')
                ?: null;
            throw new SDKError(
                message: 'Expected a response body for a non-void operation',
                status: $response->getStatusCode(),
                code: 'RESPONSE_BODY_MISSING',
                requestId: $requestId,
            );
        }

        try {
            return json_decode($contents, true, 512, JSON_THROW_ON_ERROR);
        } catch (\JsonException) {
            return $contents;
        }
    }
}

File: sdks/rust/src/api/metrics.rs
use crate::error::SdkError;
use crate::transport::Transport;
use reqwest::Method;
use serde::Serialize;
use crate::types::{ AdvisorReportResponse, EmailPerformanceResponse, EmailReadScrollAnalyticsResponse, GetMetricsAdvisorQuery, GetProjectMetricsSummaryQuery, GetProjectMetricsTimelineQuery, GetTopicPerformanceQuery, ProjectMetricsSummaryResponse, ProjectMetricsTimelineResponse, TopicPerformanceResponse };

/// MetricsAPI API operations.
pub struct MetricsAPI {
    transport: Transport,
}

impl MetricsAPI {
    pub(crate) fn new(transport: Transport) -> Self {
        Self { transport }
    }

    pub async fn get_topic_performance(
        &self,
        id: String,
        window: Option<String>,
        project_id: Option<String>,
    ) -> Result<TopicPerformanceResponse, SdkError> {
        let mut path = "/v1/metrics/topics/${id}".to_string();
        path = path.replace("${id}", &encode_path(&id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        if let Some(value) = window.as_ref() {
            query_params.push(("window".to_string(), encode_value(value)?));
        }
        if let Some(value) = project_id.as_ref() {
            query_params.push(("projectId".to_string(), encode_value(value)?));
        }
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<TopicPerformanceResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn get_email_performance(
        &self,
        id: String,
    ) -> Result<EmailPerformanceResponse, SdkError> {
        let mut path = "/v1/metrics/emails/${id}".to_string();
        path = path.replace("${id}", &encode_path(&id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<EmailPerformanceResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn get_email_engagement(
        &self,
        id: String,
    ) -> Result<EmailReadScrollAnalyticsResponse, SdkError> {
        let mut path = "/v1/metrics/emails/${id}/engagement".to_string();
        path = path.replace("${id}", &encode_path(&id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<EmailReadScrollAnalyticsResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn get_project_metrics_summary(
        &self,
        window: Option<String>,
        project_id: Option<String>,
    ) -> Result<ProjectMetricsSummaryResponse, SdkError> {
        let mut path = "/v1/metrics/summary".to_string();

        let mut query_params: Vec<(String, String)> = Vec::new();
        if let Some(value) = window.as_ref() {
            query_params.push(("window".to_string(), encode_value(value)?));
        }
        if let Some(value) = project_id.as_ref() {
            query_params.push(("projectId".to_string(), encode_value(value)?));
        }
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<ProjectMetricsSummaryResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn get_project_metrics_timeline(
        &self,
        window: Option<String>,
        granularity: Option<String>,
        format: Option<String>,
        project_id: Option<String>,
    ) -> Result<ProjectMetricsTimelineResponse, SdkError> {
        let mut path = "/v1/metrics/timeline".to_string();

        let mut query_params: Vec<(String, String)> = Vec::new();
        if let Some(value) = window.as_ref() {
            query_params.push(("window".to_string(), encode_value(value)?));
        }
        if let Some(value) = granularity.as_ref() {
            query_params.push(("granularity".to_string(), encode_value(value)?));
        }
        if let Some(value) = format.as_ref() {
            query_params.push(("format".to_string(), encode_value(value)?));
        }
        if let Some(value) = project_id.as_ref() {
            query_params.push(("projectId".to_string(), encode_value(value)?));
        }
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<ProjectMetricsTimelineResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn get_metrics_advisor(
        &self,
        project_id: Option<String>,
    ) -> Result<AdvisorReportResponse, SdkError> {
        let mut path = "/v1/metrics/advisor".to_string();

        let mut query_params: Vec<(String, String)> = Vec::new();
        if let Some(value) = project_id.as_ref() {
            query_params.push(("projectId".to_string(), encode_value(value)?));
        }
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<AdvisorReportResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

}

fn encode_path(value: &str) -> String {
    urlencoding::encode(value).into_owned()
}

fn encode_value<T: Serialize>(value: &T) -> Result<String, SdkError> {
    match serde_json::to_value(value)
        .map_err(|error| SdkError::ValidationError(error.to_string()))?
    {
        serde_json::Value::String(value) => Ok(value),
        value => Ok(value.to_string()),
    }
}

fn encode_cookie_name(name: &str) -> String {
    urlencoding::encode(name).into_owned()
}

fn encode_cookie_value<T: Serialize>(value: &T) -> Result<String, SdkError> {
    let value = encode_value(value)?;
    Ok(urlencoding::encode(&value).into_owned())
}
